common/stream-ratelimiter/src/lib.rs

```diff
@@ -76,23 +76,27 @@ impl<T: Stream> Stream for RateLimiter<T> {
     fn poll_next(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Option<Self::Item>> {
         // If the interval is ready, add permits. This is safe to do in all calls to poll_next().
         if let Poll::Ready(Some(_)) = self.as_mut().project().interval.poll_next(cx) {
-            self.permits
-                .add_permits(self.capacity - self.permits.available_permits());
+            let num_new_permits = self
+                .capacity
+                .saturating_sub(self.permits.available_permits());
+            self.permits.add_permits(num_new_permits);
```

Comment:
Okay, so this always guarantees that the `number of permits < capacity`.  Since, this is the only place we add permits, and without locking the semaphore, the number of permits is always less than or equal to the observed value.

Reply:
i'm not sure what you mean about locking the semaphore, but we actually have exclusive access in `poll_next` (since we have a `Pin<&mut Self>`), so only one thing can `poll_next` at a time. i didn't try to change any logic here, just changed the sub to saturating sub to ensure we never accidentally underflow.

as for the logic itself, it just refills the whole token bucket at every timer tick.

Reply:
Yeah, guess I was just commenting for myself.

This saturating sub is not the leak fix, but the `Arc` change is.

Reply:
err, sorry my comment above was not clear, the leak is the `ManuallyDrop`

```rust
*self.as_mut().project().permit_fut =
    Some(Box::pin(semaphore.into_permit().map(|permit| {
        // Permits should not be added back automatically. They are added on each
        // interval tick.
        ManuallyDrop::new(permit);    // <~~~~~~ this is the memory leak
    })));
```